### ISSUE TITLE
Extended-width characters only display in a single column

### ISSUE THREAD

**Issue by kkartaltepe** [mintty/mintty#615]
paste '☆' into your mintty prompt. It will appear as only left half of the character. Some other full width characters render correctly, like 你好.

The star character displays a correct full width character when I am using urxvt on linux. This happens for a handful of other characters as well but I don't have a list of all of them off hand. tested with 2.5.0 from msys2 and 2.6.2

![mintty_2016-12-15_12-50-42](https://cloud.githubusercontent.com/assets/526419/21237566/275207be-c2c5-11e6-9483-aa89d41f2054.png)

**Comment by mintty** [comment]
U+2606 WHITE STAR is a Unicode "ambiguous width" character. Displaying it wide is not *correct* but context-dependent. Ambiguous width characters are "normally" displayed narrow, i.e. using single-cell width. Only in East Asian display context their display is wide. For mintty, that would be activated by using an East Asian locale, e.g. `LC_ALL=ja_JP.UTF-8 mintty`.
The visual display, on the other hand, also depends on the font. For me, the narrow form looks good, while the wide form looks expanded. What are your settings (locale, font)? What's the locale setting when you see this character wide in urxvt?
![star](https://cloud.githubusercontent.com/assets/12740416/21242991/34282322-c317-11e6-8d9e-f678d4dbb9c8.png)
![star-wide](https://cloud.githubusercontent.com/assets/12740416/21242993/35fa8276-c317-11e6-9422-3826a794ec5a.png)

**Comment by kkartaltepe** [comment]
My msys2 system only has `LANG=en_US.UTF-8` set. LC_ALL is unset.
My urxvt system has LC_ALL is en_US.UTF-8 im not sure exactly what LANG is set to, I can get back to you with that in a couple hours.

In both systems I am using Envy Code R as the primary font. I'm fine with getting a proper single column rendering, but I just want it not to be this broken half character.

**Comment by mintty** [comment]
There is a variant Envy Code R for Powerline which has wide ambiguous characters, and maybe urxvt uses that as a criterion to switch to ambiguous wide display. That would explain the difference.
Can you try on Windows with "Envy Code R" (not "for Powerline")?

**Comment by kkartaltepe** [comment]
I am using the original official release of the font found at the creators blog
https://damieng.com/blog/2008/05/26/envy-code-r-preview-7-coding-font-released

Installed are the normal face, italic, and bold. Without the registry hack for cmd prompt and without the visual studio font hacks. I've reinstalled it for the sake of testing and it still shows the same artifact. I use the font at 10pt.

Here is the terminal using 'Courier New'
![mintty_2016-12-16_10-16-14](https://cloud.githubusercontent.com/assets/526419/21269567/cc990714-c378-11e6-8884-81caff9065dd.png)

**Comment by mintty** [comment]
The character looks "normal" for example with DejaVu Sans Mono, Everson Mono Unicode, and FreeMono.
The reason could be (not yet verified by tracing) that the character is included in those fonts but not in the others, in which case Windows applies its font fallback mechanism, turning up with a character from some other font which does not always match in cell-width. This happens to a range of character that are not typically used in terminals, like encircled numbers, Thai script characters, and lots more.

Considering what could be done about this in theory, we are facing a trade-off. Currently, via employing Uniscribe (if you use that option), extended font fallback is being used to the benefit of users displaying certain scripts. This feature should not be affected. Testing character availabilty, on the other hand, does not seem to consider font fallback. If we could check which characters would be wider than expected, we could even squeeze the rendering but at a performance penalty, I suppose...
The best workaround is to select a font that supports the characters you are using.

**Comment by mintty** [comment]
By the way, in urxvt (on Windows), the character isn't displayed properly either. It's aligned right and clipped left.

**Comment by kkartaltepe** [comment]
Sorry if you misunderstood but I meant urxvt on linux whenever I mentioned it, I dont use any other terminal emulators on windows mintty has won me over :)

**Comment by mintty** [comment]
Yes, that's why I mentioned "on Windows" explicitly. Could you make a screenshot of urxvt on Linux?

**Comment by kkartaltepe** [comment]
Unfortunately im not at that computer at the moment. Ill take a screenshot of it later tonight.

**Comment by kkartaltepe** [comment]
![virtualbox_2016-12-16_11-28-21](https://cloud.githubusercontent.com/assets/526419/21271957/d0271920-c382-11e6-8acd-483b1f7a9e45.png)
Actually got some time to spin up a vm, here is the same character rendered correctly (it looks like it is rendered as a single column character not a wide character). The font setting for Xresources used was
`URxvt.font: xft:Envy Code R:style=Regular:size=10`

**Comment by mintty** [comment]
But look at its left wing: it's a pixel shorter than the right wing, it's clipped there.

**Comment by kkartaltepe** [comment]
That is just an anti-aliasing artifact i believe, here is the same string at 20pt (increasing the pt size in mintty does not provide any more of the glyph) 
![virtualbox_2016-12-16_11-41-56](https://cloud.githubusercontent.com/assets/526419/21272424/d4547982-c384-11e6-8bcf-39abb259f1b9.png)

**Comment by mintty** [comment]
I plan to close this issue because font and width dependencies have been discussed and you can achieve the desired behaviour by starting mintty in ambiguous wide mode (corresponding to xterm -cjk_width), e.g.:
`LC_ALL=ja_JP.utf8 mintty -o Font=BatangChe`

You could be asking for an easier option to trigger this mode, without having to adjust the locale;
that would be #88.

**Comment by kkartaltepe** [comment]
I'm not sure I understand the issue linked or its significance. That looked like a proposal for a -cjk_width flag that was eventually turned down and instead opting for setting ambiguous width via a test against the Greek letter alpha's width as reported by the current font? So changing my font to one that reports Greek letter alpha's as wide would cause mintty to render ambiguous characters as wide?

Using the font in your example and setting 'Locale' in mintty options(exporting LC_ALL is overridden somewhere on my setup i guess), I do indeed get this character rendering as full width.

However, I'm not sure I see how this works in correcting(maybe "forcing" is the right word) rendering of these characters in my current font as even setting locale to ja_JP does not change character width with my chosen font (still Envy Code R). Or is this supposed to mean the only way I can change the rendering is choosing an "appropriate" font as well as an "appropriate" CJK locale.

I am on 2.7.5 if that matters.

**Comment by mintty** [comment]
> ... That looked like a proposal for a -cjk_width flag that was eventually turned down and instead opting for setting ambiguous width via a test against the Greek letter alpha's width as reported by the current font? So changing my font to one that reports Greek letter alpha's as wide would cause mintty to render ambiguous characters as wide?

That was the solution at that time, yes. But it was later changed to be triggered by locale setting rather than font properties. The font checking code is still in but I have the impression it is not effective anymore.

> Using the font in your example and setting 'Locale' in mintty options(exporting LC_ALL is overridden somewhere on my setup i guess), I do indeed get this character rendering as full width.

The advantage of using a CJK font is that the glyphs are more likely to match the expected rendering behaviour. It also works for example with DejaVu Sans Mono, but here the glyph looks more widened.
With Envy Code R actually, I cannot reproduce what you describe. There the glyph is already too wide in non-CJK locale mode, and would get widened extremely in CJK-locale-triggered ambiguous-wide mode, so that only half of the star is visible.
![615](https://cloud.githubusercontent.com/assets/12740416/26623410/ec554106-45ed-11e7-9975-7fe3347f3b5d.png)

**Comment by mintty** [comment]
The effect of the previous screenshot is actually related to font fallback. The star glyph is not contained in the font (Envy Code R), so it is taken from some other font as available (and configured) on the system.
Unfortunately, the Windows functions to determine character width (GetCharWidth*) do not consider font fallback and would report the character to be single-width although it is wide by font fallback.
There is nothing that can reasonably be done about this; in theory, questionable characters could be pre-rendered in a scratch space (not on the screen) and then measured "manually" (by checking the rendered pixels in a bitmap)...

**Comment by mintty** [comment]
Well, sometimes a theory is actually applied... Just needed to limit the performance penalty to the applicable cases.

**Comment by mintty** [comment]
Released 2.7.8.

**Comment by kkartaltepe** [comment]
Awesome just tried it out, it looks much better! Now it does look pretty normal, although like you mentioned the fallback is a different aspect ratio or something, so it cuts off somewhat like urxvt. Thanks for working on this!

**Comment by mintty** [comment]
The tweak had significant performance impact. It's limited to symbol character ranges for now (changed in repository for next release).
A better solution would be to cache detected ambiguous character withs.

**Comment by mintty** [comment]
Please recheck with 2.7.9.

**Comment by kkartaltepe** [comment]
Long delay on my end, not sure if there were other tweaks but 2.7.9 looks like it also fixed some other rendering glitches i was having with these characters. Huge thanks!

**Comment by mintty** [comment]
Thanks for the feedback; actually there was a related change meanwhile after 2.7.9. I would appreciate if you can check the current repository version too.

**Comment by kkartaltepe** [comment]
Ill give it a the repo version a go this week, since it looks like
compiling is pretty straight forward.

**Comment by kkartaltepe** [comment]
So far I have only noticed one difference between 2.7.9 and the commit that I have been running 3cabecf84a3441f4b202dfe78bed1bb7765c6e6e which had do with the ∞ glyph.
2.7.9 rendered sentence:
![2.7.9](https://i.imgur.com/eXGodrr.png)
3cabecf84a3441f4b202dfe78bed1bb7765c6e6e rendered sentence:
![master](https://i.imgur.com/hXMp6E6.png)
Sentence:
"I am erdos ∞"

**Comment by mintty** [comment]
The scale-jumping is weird and I'll try if I can tweak that.
But the main purpose of the change was to avoid clipping in as many cases as possible, especially in "ambiguous wide" mode (which people seem to appreciate in support of "Powerline fonts").
Try the following test cases (added infinity), and also with `LC_CTYPE=zh_CN.utf8 mintty`:
```
◆◇æœ━╋▅
◆H◇HæHœH━H╋H▅H
HHHHHHHHHHHHHHH
※℃℉ÅⅠ∪∬∮⑴ⓠ⓾♬♭♯∞
※H℃H℉HÅHⅠH∪H∬H∮H⑴HⓠH⓾H♬H♭H♯H∞H
```

**Comment by kkartaltepe** [comment]
Alright i found the issue (error?), I had previously in our escapades with `LC_*` types. If I have `Locale` set in my minttyrc then all LC_* types are overridden or just unset (so with `Locale=en_US` in my config any attempts at setting `LC_CTYPE` or `LC_ALL` did nothing). 

Setting `LC_CTYPE=zh_CN.utf8` the characters in your test cases are for the most part legible (without `LC_TYPE` all but the 3rd test case are illegible). However a number of the characters in `※℃℉ÅⅠ∪∬∮⑴ⓠ⓾♬♭♯∞` are single width in my fonts while some are double width. The single width characters appear to be stretched and cut off most notably Å and now that I take a closer look perhaps infinity as well)

master en_us
![master en_US](https://i.imgur.com/y7nIs59.png)
master zh_ch
![master zn_CH](https://i.imgur.com/vSACHgE.png)
2.7.9 en_us
![2.7.9](https://i.imgur.com/KwhOZNR.png)
LC_CTYPE doesnt appear to function in 2.7.9.

**Comment by kkartaltepe** [comment]
Also looks like issues with overlapping characters not re-rendering when being highlighted (from pre-2.7.9) was fixed which is nice :)

**Comment by mintty** [comment]
First, LC_CTYPE would be overriden by LC_ALL; maybe that's why it doesn't work? Use LC_ALL=.. mintty then.

I'm dismayed by your Å, I don't see that here. What's your font?
That seems to be the only glitch in your screenshot (considering that characters like encircled 10 cannot really be squeezed into a single-width cell).

**Comment by kkartaltepe** [comment]
Still using the same Envy Code R linked earlier in this thread, on windows 7 with i believe no other fonts installed. Im not to sure how to figure out where that glyph is coming from but it looks like some how the glyph with the font is not being selected. as it is listed in the gif on the fonts page and I confirmed with a font viewer that the glyph is in the font i have loaded 
![font viewer](https://i.imgur.com/fxPcxc8.png)

**Comment by mintty** [comment]
I've just installed Envy Code R (both with and without "for Powerline") and I don't see that effect.

**Comment by kkartaltepe** [comment]
I'm not sure how I can help you replicate the issue, I did notice that starting with 
`LC_CTYPE=zh_CN.utf8`
print the test to the screen,
go to options and set locale to `en_US`
the text already printed will be modified to look like
![old text after locale change](https://i.imgur.com/o8Wjfdc.png)

Not sure if this is what your text looks like or if this is just some undefined behavior. but this does make the ringed A sit nicely in its double column (similar for `♯`) and the symbol provided by the font is no longer stretched (`∪`) though `∞` which started this all does not change.

**Comment by mintty** [comment]
∞ is actually on the border of fitting into a single-width cell or not, so it might be assessed differently when scaling due to rounding. That would explain the width-jumping when zooming.
I could add a small tolerance (as I tried before) to catch that, but then other characters would be affected.
If you want to check out the algorithm for possible tweaking, it's in wininput.c: win_char_width.

**Comment by mintty** [comment]
Released 2.8.0.

**Comment by mintty** [comment]
Released 3.1.5 with explicit character width attribute support.
